contributors/design-proposals/api-machinery/admission-webhook-GA-proposal.md

```diff
@@ -0,0 +1,488 @@
+Note that namespace objects must match both object selector (if specified) and namespace selector to be sent to the Webhook except for cluster-scoped resources which namespace selector will not applied to. The propsoe change is to add an ObjectSelector to the webhook API both in v1 and v1beta1.
+
+```golang
+type Webhook struct {
+    ...
+     // ObjectSelector decides whether to run the webhook on an object based
+     // on whether the object.metadata.labels matches the selector. An object is
+     // matches both NamespaceSelector and ObjectSelector.
+     //
+     // Default to the empty LabelSelector, which matches everything.
+     // +optional
+     ObjectSelector *metav1.LabelSelector `json:"objectSelector,omitempty" protobuf:"bytes,7,opt,name=objectSelector"`
+}
+```
+
+### Scope
+
+Current webhook Rules applies to objects of all scopes. That means a Rule that uses wildcards to target an object type will be applied to both cluster scoped and namespaced objects. The proposal is to add a scope field to Admission Webhook configuration to limit webhook target on namespaced object or cluster scoped objects. The field will be added to both v1 and v1beta1. The field is optional and empty value means no scope restriction.
+
+```golang
+type ScopeType string
+
+const (
+     // ClusterScope means that scope is limited to cluster objects.
+     ClusterScope ScopeType = "Cluster"
+     // NamespacedScope means that scope is limited to namespaced objects.
+     NamespacedScope ScopeType = "Namespaced"
+)
+
+type Rule struct {
+    ...
+
+     // Scope specifies the scope of this rule. If unespecified, the scope is
+     // not limited.
+     //
+     // +optional
+     Scope ScopeType `json:"scope,omitempty" protobuf:"bytes,3,opt,name=scope"`
+}
+```
+
+### http for localhost
+
+Admission Webhook requires https for all connections. This is not necessary when the webhook in running on the same machine as apiserver as a side-car or static pod. By relaxing this restriction, it would be possible to run side-cards or static pods on the same machine without cert management. The API change for this feature is only documentation.
+
+### timeout configuration
+
+Admission Webhook has a default timeout of 30 seconds today but long running webhooks would result in a poor performance. By adding a timeout field to the configuration, the webhook author can limit the running time of the webhook that either result in failing the API call earlier or ignore the webhook call based on the failure policy. This feature, however, would not let the timeout to be extended more than todays 30 seconds default and the timeout would be defaulted to 10 seconds for v1 API while stays 30 second for v1beta API to keep backward compatibility.
```

Comment:
I'd want to think about the v1 default a little more. For now, simply note the default could be shortened in v1?

Reply:
This is the initial version of the KEP. We can change it later. What about keeping 10 for now and change it in a follow up PR. I think the KEP process encourage incremental changes to KEP instead of figuring out all details in the first PR.

Reply:
> I think the KEP process encourage incremental changes to KEP instead of figuring out all details in the first PR.

Indicating which details are known to be less certain is helpful for readers